wrappers/dotnet/indy-sdk-dotnet-test/Wrapper/AgentTests/AgentSendTest.cs

```diff
@@ -1,69 +1,45 @@
-        private static TaskCompletionSource<string> clientToServerMsgFuture = new TaskCompletionSource<string>();
-
-        private new static MessageReceivedHandler _messageObserver = (Connection connection, string message) =>
-            {
-                Console.WriteLine("Received message '" + message + "' on connection " + connection);
-
-                serverToClientMsgFuture.SetResult(message);
-            };
-
-        private new static MessageReceivedHandler _messageObserverForIncoming = (Connection connection, string message) =>
-            {
-                Console.WriteLine("Received message '" + message + "' on incoming connection " + connection);
-
-                clientToServerMsgFuture.SetResult(message);
-            };
-
-        private new static ConnectionOpenedHandler _incomingConnectionObserver = (Listener listener, Connection connection, string senderDid, string receiverDid) =>
-            {
-                Console.WriteLine("New connection " + connection);
-
-                serverToClientConnectionFuture.SetResult(connection);
-
-                return _messageObserverForIncoming;
-            };
-
         [TestMethod]
-        public void TestAgentSendWorks()
+        public async Task TestAgentSendWorks()
         {
             var endpoint = "127.0.0.1:9609";
 
-            var myDidResult = Signus.CreateAndStoreMyDidAsync(_wallet, "{}").Result;
+            var myDidResult = await Signus.CreateAndStoreMyDidAsync(_wallet, "{}");
 
             var identityJson = string.Format("{{\"did\":\"{0}\", \"pk\":\"{1}\", \"verkey\":\"{2}\", \"endpoint\":\"{3}\"}}",
                     myDidResult.Did, myDidResult.Pk, myDidResult.VerKey, endpoint);
-            Signus.StoreTheirDidAsync(_wallet, identityJson).Wait();
+            await Signus.StoreTheirDidAsync(_wallet, identityJson);
 
-            var activeListener = Agent.AgentListenAsync(endpoint, _incomingConnectionObserver).Result;
+            var listener = await Agent.AgentListenAsync(endpoint);
+            await listener.AddIdentityAsync(_pool, _wallet, myDidResult.Did);
 
-            activeListener.AddIdentityAsync(_pool, _wallet, myDidResult.Did).Wait();
+            var clientConnection = await Agent.AgentConnectAsync(_pool, _wallet, myDidResult.Did, myDidResult.Did);
 
-            var clientToServerConnection = Agent.AgentConnectAsync(_pool, _wallet, myDidResult.Did, myDidResult.Did, _messageObserver).Result;
+            var connectionEvent = await listener.WaitForConnection();
+            var serverConnection = connectionEvent.Connection;
 
             var clientToServerMessage = "msg_from_client";
             var serverToClientMessage = "msg_from_server";
 
-            clientToServerConnection.SendAsync(clientToServerMessage).Wait();
+            await clientConnection.SendAsync(clientToServerMessage);
+
+            var serverMessageEvent = await serverConnection.WaitForMessage();
```

Comment:
Yes, I suggest trying rewriting of this tests with WaitForAny usage and check how usable it will be.